[channel-manager] channel selection

abtink opened the pull request:
There are 3 commits in this PR:

 **[channel-manager] add (auto) "channel selection" algorithm**   
This commit updates `ChannelManager` class to add support for "channel selection" logic where the `ChannelManager` can select a channel (for network to switch to) based on the collected channel/link quality data (by `ChannelMonitor` module). This commit also adds support for auto-channel-selection functionality where `ChannelManager` will periodically check to see if network can/should switch to a better channel. This commit also adds public OpenThread APIs for all newly added functionalities.

**[ncp] implement "Channel Manager" spinel properties**
This commit adds new spinel properties along with their set/get
handlers for `ChannelManager` module.

**[tests] add new toranj test scripts for "ChannelManager" module**
This commit adds new `toranj` test scripts to test newly added functionality by `ChannelManager` including:    
- Test to check new spinel/wpan properties,
- Test to check channel change request feature,
- Test to check channel selection feature.  

This commit also updates the posix/radio `otPlatRadioGetRssi()` to emulate a simple interference model where it would return either a high or a low  RSSI value with a fixed probability per each channel. This is used to verify the `ChannelMonitor` sampling logic and channel selection functionality.

jwhui commented on include/openthread/channel_manager.h:
```diff
@@ -104,6 +101,82 @@ uint16_t otChannelManagerGetDelay(otInstance *aInstance);
 otError otChannelManagerSetDelay(otInstance *aInstance, uint16_t aDelay);
 
+/**
+ * This functions requests that `ChannelManager` checks and selects a new channel and starts a channel change
```
nit: "functions" -> "function"
nit: missing period at end

jwhui commented on src/core/openthread-core-default-config.h:
```diff
@@ -1102,6 +1127,78 @@
+/**
+ * @def OPENTHREAD_CONFIG_CHANNEL_MANAGER_DEFAULT_AUTO_SELECT_INTERVAL
+ *
+ * The default period interval in seconds used Channel Manager for auto-channel-selection functionality
```
nit: "used Channel" -> "used by Channel"

jwhui commented on src/core/utils/channel_manager.cpp:
```diff
@@ -273,14 +267,247 @@ void ChannelManager::HandleStateChanged(uint32_t aFlags)
+    // the best overall by at least `kQualityThresholdToSkipFavored`.
+
+    if (favoredBest.IsEmpty() || ((favoredQuality >= kQualityThresholdToSkipFavored) &&
+                                  (supportedQuality < favoredQuality - kQualityThresholdToSkipFavored)))
```
Add some preprocessor checks to the config so that we do not need to worry about underflow here?

jwhui commented on src/core/utils/channel_manager.hpp:
```diff
@@ -91,13 +92,12 @@ class ChannelManager : public InstanceLocator
      * A subsequent call to this method will cancel an ongoing previously requested channel change.
      *
-     * @param[in] aChannel             The new channel for the Thread network.
+     * If the requested channel changes, it will trigger  a `Notifier` event `OT_CHANGED_CHANNEL_MANAGER_NEW_CHANNEL`.
```
nit: double space in "trigger  a"

jwhui commented on src/core/utils/channel_manager.hpp:
```diff
@@ -129,33 +129,138 @@ class ChannelManager : public InstanceLocator
+    otError RequestChannelSelect(bool aSkipQualityCheck);
+
+    /**
+     * This method enables/disabled the auto-channel-selection functionality.
```
nit: "disabled" -> "disables"

jwhui commented on src/core/utils/channel_manager.hpp:
```diff
@@ -129,33 +129,138 @@ class ChannelManager : public InstanceLocator
+     * @returns TRUE if enabled, FALSE otherwise.
+     *
+     */
+    bool GetAutoChannelSelectionEnabled(void) { return mAutoSelectEnabled; }
```
Declare as `const`.

jwhui commented on src/core/utils/channel_manager.hpp:
```diff
@@ -129,33 +129,138 @@ class ChannelManager : public InstanceLocator
+     * @returns The interval in seconds.
+     *
+     */
+    uint32_t GetAutoChannelSelectionInterval(void) { return mAutoSelectInterval; }
```
Declare as `const`.

jwhui commented on src/core/utils/channel_monitor.cpp:
```diff
@@ -213,6 +213,55 @@ void ChannelMonitor::LogResults(void)
+
+        if (bestMask.IsEmpty() || (quality <= aBestQuality))
+        {
+            if (quality < aBestQuality)
```
It's a bit confusing to have lower integer values actually have higher quality.  I would consider renaming or changing the polarity.

jwhui commented on src/core/utils/channel_monitor.hpp:
```diff
@@ -168,6 +169,21 @@ class ChannelMonitor : public InstanceLocator
+     * higher than `kRssiThreshold`) and lower value (less interfered) is considered better.
+     *
+     * @param[in]  aMask         A channel mask (the search is limited to channels in @p aMask).
+     * @param[out] aQuality      A reference to `uint16` to return the quality value associated with best channel(s).
```
Larger values is really lower quality, right?  Can we make the PR more intuitive and consistent?  For example, switch everything to "busy", "utilization", or "occupancy"?

abtink replied:
Will go for "occupancy". Probably will add a new commit (before the existing 3) to change this.

abtink commented:
In new push, added a new commit `[channel-monitor] use "channel occupancy" instead of "quality". The other commits are also changed to use "occupancy" term.
